File: apps/ccv-service/src/models/ccv.py
"""CCV data models."""

from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


class TermStatus(str, Enum):
    """Status of a CCV term."""
    DRAFT = "draft"           # Newly suggested, not reviewed
    PENDING = "pending"       # Awaiting approval
    APPROVED = "approved"     # Approved and active
    DEPRECATED = "deprecated" # No longer in use
    REJECTED = "rejected"     # Rejected suggestion


class SuggestionType(str, Enum):
    """Type of CCV suggestion."""
    NEW_TERM = "new_term"           # Suggest a new canonical term
    SYNONYM = "synonym"             # Suggest a synonym mapping
    HIERARCHY = "hierarchy"         # Suggest parent-child relationship
    MERGE = "merge"                 # Suggest merging two terms
    DEFINITION = "definition"       # Suggest/update definition
    DEPRECATION = "deprecation"     # Suggest deprecating a term


class SuggestionStatus(str, Enum):
    """Status of a suggestion."""
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    AUTO_APPROVED = "auto_approved"


# =============================================================================
# Core Term Models
# =============================================================================

class CCVTerm(BaseModel):
    """A canonical control vocabulary term."""
    id: UUID = Field(default_factory=uuid4)
    canonical_name: str
    definition: str | None = None
    domain: str | None = None  # e.g., "finance", "security", "hr"
    parent_id: UUID | None = None  # For hierarchy
    status: TermStatus = TermStatus.APPROVED

    # Metadata
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    created_by: UUID | None = None
    approved_by: UUID | None = None
    approved_at: datetime | None = None

    # Source tracking
    source: str | None = None  # "manual", "auto_extracted", "imported"
    confidence: float = 1.0  # For auto-extracted terms

    # Usage statistics
    usage_count: int = 0
    last_used_at: datetime | None = None


class CCVSynonym(BaseModel):
    """A synonym or alias for a CCV term."""
    id: UUID = Field(default_factory=uuid4)
    term_id: UUID  # The canonical term this is a synonym for
    synonym: str
    source: str | None = None  # Where this synonym was found
    data_source_id: str | None = None  # Which data source uses this variant
    confidence: float = 1.0
    status: TermStatus = TermStatus.APPROVED
    created_at: datetime = Field(default_factory=datetime.utcnow)
    created_by: UUID | None = None


class CCVRelationship(BaseModel):
    """A relationship between CCV terms (beyond parent-child)."""
    id: UUID = Field(default_factory=uuid4)
    source_term_id: UUID
    target_term_id: UUID
    relationship_type: str  # "related_to", "see_also", "broader_than", "narrower_than"
    confidence: float = 1.0
    created_at: datetime = Field(default_factory=datetime.utcnow)
    created_by: UUID | None = None


# =============================================================================
# Suggestion Models
# =============================================================================

class CCVSuggestion(BaseModel):
    """A suggestion for CCV update."""
    id: UUID = Field(default_factory=uuid4)
    suggestion_type: SuggestionType
    status: SuggestionStatus = SuggestionStatus.PENDING
    confidence: float  # LLM confidence score

    # What's being suggested
    suggested_value: str
    context: str | None = None  # The context where this was found

    # Related entities
    term_id: UUID | None = None  # Existing term (for synonyms, hierarchy, etc.)
    related_term_id: UUID | None = None  # For merges, relationships
    parent_term_id: UUID | None = None  # For hierarchy suggestions

    # Source tracking
    source_entity_id: UUID | None = None  # Graph entity that triggered this
    source_data_source: str | None = None
    source_text: str | None = None  # Original text that was analyzed

    # LLM reasoning
    llm_reasoning: str | None = None
    alternative_suggestions: list[str] = Field(default_factory=list)

    # Metadata
    created_at: datetime = Field(default_factory=datetime.utcnow)
    reviewed_at: datetime | None = None
    reviewed_by: UUID | None = None
    review_notes: str | None = None


class DataSourceMapping(BaseModel):
    """Maps a data source field value to a CCV term."""
    id: UUID = Field(default_factory=uuid4)
    term_id: UUID
    data_source_id: str
    original_value: str  # The value as it appears in the source
    field_path: str | None = None  # e.g., "employee.department"
    confidence: float = 1.0
    is_auto_mapped: bool = False
    created_at: datetime = Field(default_factory=datetime.utcnow)


# =============================================================================
# API Request/Response Models
# =============================================================================

class TermCreate(BaseModel):
    """Request to create a new term."""
    canonical_name: str
    definition: str | None = None
    domain: str | None = None
    parent_id: UUID | None = None
    synonyms: list[str] = Field(default_factory=list)


class TermUpdate(BaseModel):
    """Request to update a term."""
    canonical_name: str | None = None
    definition: str | None = None
    domain: str | None = None
    parent_id: UUID | None = None
    status: TermStatus | None = None


class SynonymCreate(BaseModel):
    """Request to add a synonym."""
    synonym: str
    data_source_id: str | None = None


class SuggestionReview(BaseModel):
    """Request to review a suggestion."""
    action: str  # "approve", "reject", "modify"
    modified_value: str | None = None  # If action is "modify"
    notes: str | None = None


class TermWithDetails(BaseModel):
    """Term with all related data."""
    term: CCVTerm
    synonyms: list[CCVSynonym] = Field(default_factory=list)
    children: list["TermWithDetails"] = Field(default_factory=list)
    parent: CCVTerm | None = None
    relationships: list[CCVRelationship] = Field(default_factory=list)
    mappings: list[DataSourceMapping] = Field(default_factory=list)
    pending_suggestions: int = 0


class HierarchyNode(BaseModel):
    """A node in the term hierarchy tree."""
    term: CCVTerm
    children: list["HierarchyNode"] = Field(default_factory=list)
    synonym_count: int = 0
    usage_count: int = 0


class TermSearchResult(BaseModel):
    """Search result for terms."""
    term: CCVTerm
    match_type: str  # "canonical", "synonym", "definition"
    matched_text: str
    score: float


# =============================================================================
# Event Models (for real-time updates)
# =============================================================================

class CCVEvent(BaseModel):
    """Event for real-time CCV updates."""
    event_type: str  # "term_created", "term_updated", "suggestion_created", etc.
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    data: dict[str, Any]
    user_id: UUID | None = None


# Enable forward references
TermWithDetails.model_rebuild()
HierarchyNode.model_rebuild()

File: apps/web/src/lib/api/ccv.ts
/**
 * CCV API client for interacting with the CCV service through Next.js proxy routes.
 */

export interface CCVTerm {
  id: string;
  canonical_name: string;
  definition: string | null;
  domain: string | null;
  parent_id: string | null;
  status: 'pending' | 'approved' | 'deprecated' | 'rejected';
  source: string | null;
  confidence: number;
  usage_count: number;
  created_at: string;
  updated_at: string;
}

export interface CCVSynonym {
  id: string;
  term_id: string;
  synonym: string;
  source: string | null;
  data_source_id: string | null;
  confidence: number;
  status: string;
  created_at: string;
}

export interface CCVSuggestion {
  id: string;
  suggestion_type: 'new_term' | 'synonym' | 'hierarchy' | 'merge' | 'definition';
  status: 'pending' | 'approved' | 'rejected' | 'auto_approved';
  confidence: number;
  suggested_value: string;
  context: string | null;
  term_id: string | null;
  parent_term_id: string | null;
  source_entity_id: string | null;
  source_data_source: string | null;
  source_text: string | null;
  llm_reasoning: string | null;
  alternative_suggestions: string[];
  created_at: string;
  reviewed_at: string | null;
  reviewed_by: string | null;
  review_notes: string | null;
}

export interface HierarchyNode {
  term: CCVTerm;
  children: HierarchyNode[];
  synonym_count: number;
  usage_count: number;
}

export interface TermWithDetails extends CCVTerm {
  synonyms: CCVSynonym[];
  children: CCVTerm[];
  parent: CCVTerm | null;
}

// Terms API
export async function listTerms(params?: {
  domain?: string;
  status?: string;
  page?: number;
  page_size?: number;
}): Promise<CCVTerm[]> {
  const searchParams = new URLSearchParams();
  if (params?.domain) searchParams.set('domain', params.domain);
  if (params?.status) searchParams.set('status', params.status);
  if (params?.page) searchParams.set('page', params.page.toString());
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString());

  const response = await fetch(`/api/ccv/terms?${searchParams}`);
  if (!response.ok) throw new Error('Failed to fetch terms');
  return response.json();
}

export async function getTerm(termId: string): Promise<TermWithDetails> {
  const response = await fetch(`/api/ccv/terms/${termId}`);
  if (!response.ok) throw new Error('Failed to fetch term');
  return response.json();
}

export async function createTerm(data: {
  canonical_name: string;
  definition?: string;
  domain?: string;
  parent_id?: string;
}): Promise<CCVTerm> {
  const response = await fetch('/api/ccv/terms', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!response.ok) throw new Error('Failed to create term');
  return response.json();
}

export async function updateTerm(
  termId: string,
  data: Partial<{
    canonical_name: string;
    definition: string;
    domain: string;
    parent_id: string | null;
    status: string;
  }>
): Promise<CCVTerm> {
  const response = await fetch(`/api/ccv/terms/${termId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!response.ok) throw new Error('Failed to update term');
  return response.json();
}

export async function deleteTerm(termId: string): Promise<void> {
  const response = await fetch(`/api/ccv/terms/${termId}`, {
    method: 'DELETE',
  });
  if (!response.ok) throw new Error('Failed to delete term');
}

export async function searchTerms(query: string, domain?: string): Promise<CCVTerm[]> {
  const searchParams = new URLSearchParams({ q: query });
  if (domain) searchParams.set('domain', domain);

  const response = await fetch(`/api/ccv/terms/search?${searchParams}`);
  if (!response.ok) throw new Error('Failed to search terms');
  return response.json();
}

// Synonyms API
export async function addSynonym(
  termId: string,
  synonym: string
): Promise<CCVSynonym> {
  const response = await fetch(`/api/ccv/terms/${termId}/synonyms`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ synonym }),
  });
  if (!response.ok) throw new Error('Failed to add synonym');
  return response.json();
}

export async function removeSynonym(termId: string, synonymId: string): Promise<void> {
  const response = await fetch(
    `/api/ccv/terms/${termId}/synonyms/${synonymId}`,
    { method: 'DELETE' }
  );
  if (!response.ok) throw new Error('Failed to remove synonym');
}

// Suggestions API
export async function listSuggestions(params?: {
  status?: string;
  suggestion_type?: string;
  page?: number;
  page_size?: number;
}): Promise<CCVSuggestion[]> {
  const searchParams = new URLSearchParams();
  if (params?.status) searchParams.set('status', params.status);
  if (params?.suggestion_type) searchParams.set('suggestion_type', params.suggestion_type);
  if (params?.page) searchParams.set('page', params.page.toString());
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString());

  const response = await fetch(`/api/ccv/suggestions?${searchParams}`);
  if (!response.ok) throw new Error('Failed to fetch suggestions');
  return response.json();
}

export async function getPendingSuggestions(): Promise<CCVSuggestion[]> {
  const response = await fetch('/api/ccv/suggestions/pending');
  if (!response.ok) throw new Error('Failed to fetch pending suggestions');
  return response.json();
}

export async function reviewSuggestion(
  suggestionId: string,
  action: 'approve' | 'reject',
  notes?: string,
  modified_value?: string
): Promise<CCVSuggestion> {
  const response = await fetch(`/api/ccv/suggestions/${suggestionId}/review`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ action, notes, modified_value }),
  });
  if (!response.ok) throw new Error('Failed to review suggestion');
  return response.json();
}

export async function batchReviewSuggestions(
  suggestionIds: string[],
  action: 'approve' | 'reject'
): Promise<{ reviewed: number; total: number }> {
  const response = await fetch('/api/ccv/suggestions/batch-review', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ suggestion_ids: suggestionIds, action }),
  });
  if (!response.ok) throw new Error('Failed to batch review suggestions');
  return response.json();
}

// Hierarchy API
export async function getHierarchy(params?: {
  domain?: string;
  max_depth?: number;
}): Promise<HierarchyNode[]> {
  const searchParams = new URLSearchParams();
  if (params?.domain) searchParams.set('domain', params.domain);
  if (params?.max_depth) searchParams.set('max_depth', params.max_depth.toString());

  const response = await fetch(`/api/ccv/hierarchy?${searchParams}`);
  if (!response.ok) throw new Error('Failed to fetch hierarchy');
  return response.json();
}

export async function getHierarchyStats(): Promise<{
  total_terms: number;
  root_terms: number;
  max_depth: number;
  domains: string[];
}> {
  const response = await fetch('/api/ccv/hierarchy/stats');
  if (!response.ok) throw new Error('Failed to fetch hierarchy stats');
  return response.json();
}

export async function moveTerm(termId: string, newParentId: string | null): Promise<CCVTerm> {
  const response = await fetch(`/api/ccv/hierarchy/${termId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ new_parent_id: newParentId }),
  });
  if (!response.ok) throw new Error('Failed to move term');
  return response.json();
}

export async function getAncestors(termId: string): Promise<CCVTerm[]> {
  const response = await fetch(`/api/ccv/hierarchy/${termId}/ancestors`);
  if (!response.ok) throw new Error('Failed to fetch ancestors');
  return response.json();
}

export async function getDescendants(termId: string): Promise<CCVTerm[]> {
  const response = await fetch(`/api/ccv/hierarchy/${termId}/descendants`);
  if (!response.ok) throw new Error('Failed to fetch descendants');
  return response.json();
}

// Extraction API
export async function extractTermsFromText(data: {
  text: string;
  context?: string;
  domain?: string;
}): Promise<{ suggestion_count: number; suggestions: CCVSuggestion[] }> {
  const response = await fetch('/api/ccv/extraction/text', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!response.ok) throw new Error('Failed to extract terms');
  return response.json();
}

export async function analyzeVocabularyCoverage(data: {
  text: string;
  domain?: string;
}): Promise<{
  total_unique_words: number;
  matched_terms: number;
  unmatched_words: number;
  coverage_percentage: number;
  matched_examples: string[];
  unmatched_examples: string[];
}> {
  const response = await fetch('/api/ccv/extraction/analyze', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!response.ok) throw new Error('Failed to analyze coverage');
  return response.json();
}
